iroh-qrpc-types/src/p2p.rs

```diff
@@ -0,0 +1,482 @@
+    type Req = P2pRequest;
+    type Res = P2pResponse;
+}
+
+impl RpcMsg<P2pService> for VersionRequest {
+    type Response = VersionResponse;
+}
+
+impl RpcMsg<P2pService> for ShutdownRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for BitswapRequest {
+    type Response = RpcResult<BitswapResponse>;
+}
+
+impl Msg<P2pService> for FetchProvidersDhtRequest {
+    type Response = RpcResult<FetchProvidersDhtResponse>;
+
+    type Update = Self;
+
+    type Pattern = ServerStreaming;
+}
+
+impl RpcMsg<P2pService> for StopSessionBitswapRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for NotifyNewBlocksBitswapRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for GetListeningAddrsRequest {
+    type Response = RpcResult<GetListeningAddrsResponse>;
+}
+
+impl RpcMsg<P2pService> for GetPeersRequest {
+    type Response = RpcResult<GetPeersResponse>;
+}
+
+impl RpcMsg<P2pService> for ConnectRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for ConnectByPeerIdRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for DisconnectRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for LookupRequest {
+    type Response = RpcResult<LookupResponse>;
+}
+
+impl RpcMsg<P2pService> for LookupLocalRequest {
+    type Response = RpcResult<LookupResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubAddExplicitPeerRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for GossipsubAllMeshPeersRequest {
+    type Response = RpcResult<GossipsubPeersResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubMeshPeersRequest {
+    type Response = RpcResult<GossipsubPeersResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubAllPeersRequest {
+    type Response = RpcResult<GossipsubAllPeersResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubPublishRequest {
+    type Response = RpcResult<GossipsubPublishResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubTopicsRequest {
+    type Response = RpcResult<GossipsubTopicsResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubSubscribeRequest {
+    type Response = RpcResult<GossipsubSubscribeResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubUnsubscribeRequest {
+    type Response = RpcResult<GossipsubUnsubscribeResponse>;
+}
+
+impl RpcMsg<P2pService> for GossipsubRemoveExplicitPeerRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for StartProvidingRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for StopProvidingRequest {
+    type Response = RpcResult<()>;
+}
+
+impl RpcMsg<P2pService> for LocalPeerIdRequest {
+    type Response = RpcResult<LocalPeerIdResponse>;
+}
+
+impl RpcMsg<P2pService> for ExternalAddrsRequest {
+    type Response = RpcResult<ExternalAddrsResponse>;
+}
+
+impl RpcMsg<P2pService> for ListenersRequest {
+    type Response = RpcResult<ListenersResponse>;
+}
```

Comment:
is the above covered by the macros? this is pretty painful to look at

Reply:
Yes, the macro written by @Frando allows you to declare these.

But I don't mind it that much, compared to a giant macro machine that is kinda write only.

It really sucks if you go into a foreign code base, use go to definition, and then just get pointed into a single line macro invocation all the time. (That was basically me when I tried to add client streams to the existing RPC stuff before ipfs camp)

Reply:
It's pretty boring to write, but boring is good.  This way makes it easy to jump around to the right places.

Reply:
This is the part which the macro also does fine, if you want things a bit more compact. Where I am really not so sure is the dispatcher. There I really think the verbosity is preferable to hiding details.